### PR TITLE
fix: organisation bug,owner feature ,  notification improvemnt

### PR DESCRIPTION
**Pull request by Sad-lie**
# Pull Request Description

## Changes Made

- [x] Feature addition
- [x] Bug fix

## Description

minor fixes, added owner , notification for documents

## Motivation and Context

to make possible to diffrentiate owner , added notification for documents

## How Has This Been Tested?

via terminal and frontend

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My changes generate no new warnings.
- [ ] I have checked my code and corrected any misspellings.

### CODE REVIEW

File: lib/wraft_doc_web/controllers/organisation_controller.ex [hunk 1]
```diff
@@ -635,6 +635,18 @@ defmodule WraftDocWeb.Api.V1.OrganisationController do
     response(401, "Unauthorized", Schema.ref(:Error))
   end
 
+  def organisation_transfer(conn, %{"id" => new_user_id}) do
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/organisation_controller.ex:638]: Change function name to `transfer_ownership`.

File: lib/wraft_doc_web/controllers/organisation_controller.ex [hunk 2]
```diff
@@ -635,6 +635,18 @@ defmodule WraftDocWeb.Api.V1.OrganisationController do
...
+    current_user = conn.assigns[:current_user]
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/organisation_controller.ex:639]:
```elixir
    %{current_org_id: organisation_id} = current_user = conn.assigns[:current_user]
```

line 642, 
```elixir
Enterprise.get_organisation(organisation_id),
```

File: lib/wraft_doc_web/controllers/organisation_controller.ex [hunk 3]
```diff
@@ -635,6 +635,18 @@ defmodule WraftDocWeb.Api.V1.OrganisationController do
...
+
+    with %Organisation{} = organisation <-
+           Enterprise.get_organisation(current_user[:current_org]),
+         new_user <- Account.get_user(new_user_id),
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/organisation_controller.ex:643]: Account.get_user is not required  instead can use , ensure to add pattern matching in the output

```elixir
@doc """
   Gets the UserOrganisation for given user ID and organisation ID
  """
  @spec get_user_organisation(User.t(), Ecto.UUID.t()) :: UserOrganisation.t() | nil
  def get_user_organisation(%User{current_org_id: org_id}, user_id) do
    Repo.get_by(UserOrganisation, organisation_id: org_id, user_id: user_id)
  end
  ```

  Enterprise.ex:1796

File: lib/wraft_doc/organisation/enterprise.ex [hunk 4]
```diff
@@ -816,6 +818,19 @@ defmodule WraftDoc.Enterprise do
     |> Oban.insert()
   end
 
+  def transfer_ownership(
+        %Organisation{owner_id: owner_id} = organisation,
+        %User{id: owner_id} = _current_owner,
+        %User{id: new_owner_id} = _new_owner
+      ) do
+    organisation
+    |> Ecto.Changeset.change(owner_id: new_owner_id)
```
Comment (vasil1729) [inline lib/wraft_doc/organisation/enterprise.ex:827]:
```elixir
  def update_owner_changeset(%Organisation{} = organisation, attrs \\ %{}) do
    organisation
    |> cast(attrs, [:owner_id])
    |> validate_required([:owner_id])
  end
  ```

  Add this to the `organisation.ex`

  ```elixir
  def transfer_ownership(organisation, new_owner) do
    organisation
    |> Organisation.update_owner_changeset(%{owner_id: new_owner.id})
    |> Repo.update()
```

File: lib/wraft_doc/organisation/enterprise.ex [hunk 5]
```diff
@@ -816,6 +818,19 @@ defmodule WraftDoc.Enterprise do
...
+    |> Repo.update()
+  end
+
+  def transfer_ownership(%Organisation{}, %User{}, %User{}),
```
Comment (vasil1729) [inline lib/wraft_doc/organisation/enterprise.ex:831]: Remove this as the check would be done via pattern matching from the controller action itself.

File: lib/wraft_doc_web/controllers/organisation_controller.ex [hunk 6]
```diff
@@ -635,6 +635,18 @@ defmodule WraftDocWeb.Api.V1.OrganisationController do
     response(401, "Unauthorized", Schema.ref(:Error))
   end
 
+  def organisation_transfer(conn, %{"id" => new_user_id}) do
+    current_user = conn.assigns[:current_user]
+
+    with %Organisation{} = organisation <-
+           Enterprise.get_organisation(current_user[:current_org]),
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/organisation_controller.ex:642]:
```elixir
  %{current_org_id: organisation_id, id: user_id} = current_user = conn.assigns[:current_user]

    with %Organisation{owner_id: ^user_id} = organisation <-
```

Add the above pinoperator , to check if the current user is the current owner.

Test this scenario for both the success case and failure case and ensure that the error handling is returned gracefully.

File: priv/repo/migrations/20250214083221_add_owner_to_organisation.exs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+defmodule WraftDoc.Repo.Migrations.AddOwnerToOrganisation do
+  use Ecto.Migration
+
+  def change do
+    alter table(:organisation) do
+      add(:owner_id, :uuid)
```
Comment (vasil1729) [inline priv/repo/migrations/20250214083221_add_owner_to_organisation.exs:6]:
```elixir
    add(:owner_id, references(:user, type: :uuid, on_delete: :nilify_all))
```

File: lib/wraft_doc_web/controllers/organisation_controller.ex [hunk 8]
```diff
@@ -635,6 +635,18 @@ defmodule WraftDocWeb.Api.V1.OrganisationController do
...
+         new_user <- Account.get_user(new_user_id),
+         {:ok, %Organisation{}} <-
+           Enterprise.transfer_ownership(organisation, current_user, new_user) do
+      render(conn, "create.json")
+    end
+  end
+
   @doc """
     Remove a user from the organisation
   """
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/organisation_controller.ex:652]: Ensure that with remove_user api, user is not allowed to leave the organisation without transferring ownership. 

So basically ensure if the user to be removed from the organisation is not the owner of the organisation.

File: lib/wraft_doc/notifications/notifications.ex [hunk 9]
```diff
@@ -105,6 +105,33 @@ defmodule WraftDoc.Notifications do
     })
   end
 
+  @doc """
+  Notification for the Documet flow
+  """
+  def document_notification(current_user, instance, organisation, state) do
```
Comment (vasil1729) [inline lib/wraft_doc/notifications/notifications.ex:111]: Use pattern matching to add values

File: lib/wraft_doc/document/document.ex [hunk 10]
```diff
@@ -849,6 +849,10 @@ defmodule WraftDoc.Document do
     Repo.one(query) || current_state.id
   end
 
+  def next_state(currrent_state) do
+    Repo.preload(Repo.get(State, next_state_id(currrent_state)), :approvers)
```
Comment (vasil1729) [inline lib/wraft_doc/document/document.ex:853]: Refactor and write using pipe operator

File: lib/wraft_doc/document/document.ex [hunk 11]
```diff
@@ -849,6 +849,10 @@ defmodule WraftDoc.Document do
     Repo.one(query) || current_state.id
   end
```
Comment (vasil1729) [inline lib/wraft_doc/document/document.ex:851]: Add spec and docs for next_state function

File: lib/wraft_doc_web/controllers/instance_controller.ex [hunk 12]
```diff
@@ -973,11 +973,23 @@ defmodule WraftDocWeb.Api.V1.InstanceController do
     %{current_org_id: organisation_id} = current_user = conn.assigns.current_user
 
     with %Instance{
-           content_type: %ContentType{organisation_id: ^organisation_id},
-           state: _state
+           content_type: %ContentType{
+             organisation_id: ^organisation_id,
+             organisation: organisation
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/instance_controller.ex:978]:
```elixir
organisation: %Organisation{id: ^organisation_id} = organisation
```

File: lib/wraft_doc/notifications/notifications.ex [hunk 13]
```diff
@@ -105,6 +105,33 @@ defmodule WraftDoc.Notifications do
...
+    end)
+    |> Multi.run(:pending_approvals_notification, fn _repo, _changes ->
+      create_notification(new_state.approvers, %{
+        type: :pending_approvals,
+        document_title: instance.serialized["title"],
+        organisation_name: organisation.name,
+        state_name: new_state.state
+      })
+    end)
+    |> Repo.transaction()
```
Comment (vasil1729) [inline lib/wraft_doc/notifications/notifications.ex:132]: Instead of firing a single notification from here, fire or dispatch both the notification asynchronously from controller action, keep the logic simple here.

File: lib/wraft_doc/organisation/enterprise.ex [hunk 14]
```diff
@@ -692,10 +692,12 @@ defmodule WraftDoc.Enterprise do
   """
   @spec create_organisation(User.t(), map) :: Organisation.t()
   def create_organisation(%User{} = user, params) do
+    updated_params = Map.put(params, "owner_id", user.id)
```
Comment (vasil1729) [inline lib/wraft_doc/organisation/enterprise.ex:695]:
```elixir
def create_organisation(%User{id: user_id} = user, params) do

params = Map.merge(params, %{"owner_id" => user_id})
```

File: lib/wraft_doc/documents/documents.ex [hunk 15]
```diff
@@ -317,6 +317,16 @@ defmodule WraftDoc.Documents do
     Repo.one(query) || current_state.id
   end
 
+  @doc """
+    gives next state of flow
```
Comment (vasil1729) [inline lib/wraft_doc/documents/documents.ex:321]: @doc """
   Next state of the document flow.
"""

File: lib/wraft_doc/notifications/notifications.ex [hunk 16]
```diff
@@ -105,6 +107,38 @@ defmodule WraftDoc.Notifications do
     })
   end
 
+  @doc """
+  Notification for the Documet flow
```
Comment (vasil1729) [inline lib/wraft_doc/notifications/notifications.ex:111]: @doc """
Notification for the document flow.
"""

File: lib/wraft_doc/notifications/notifications.ex [hunk 17]
```diff
@@ -105,6 +107,38 @@ defmodule WraftDoc.Notifications do
...
+  """
+  def document_notification(
+        %User{name: approver_name} = _current_user,
+        %Instance{serialized: %{"title" => document_title}} = _instance,
+        %Organisation{name: organisation_name} = _organisation,
+        state
+      ) do
+    new_state = Documents.next_state(state)
```
Comment (vasil1729) [inline lib/wraft_doc/notifications/notifications.ex:119]:
```elixir
next_state = Documents.next_state(state)
```

Test for the edge cases where the next state is nil.

File: lib/wraft_doc/organisation/enterprise.ex [hunk 18]
```diff
@@ -1770,20 +1785,49 @@ defmodule WraftDoc.Enterprise do
...
-    |> Repo.update()
+  def remove_user(%UserOrganisation{user: user} = user_organisation, org_id) do
+    with :ok <- handle_last_signed_in_org(user, org_id) do
+      user_organisation
+      |> UserOrganisation.delete_changeset(%{deleted_at: NaiveDateTime.local_now()})
+      |> Repo.update()
+    end
+  end
+
+  def validate_owner_removal(user_id, owner_id) do
```
Comment (vasil1729) [inline lib/wraft_doc/organisation/enterprise.ex:1796]: Add spec and docs for `validate_owner_removal` function

File: lib/wraft_doc/organisation/enterprise.ex [hunk 19]
```diff
@@ -1770,20 +1785,49 @@ defmodule WraftDoc.Enterprise do
...
+    case user_id == owner_id do
+      true -> {:error, :cannot_remove_owner}
+      false -> :ok
+    end
+  end
+
+  def handle_last_signed_in_org(user, org_id) do
```
Comment (vasil1729) [inline lib/wraft_doc/organisation/enterprise.ex:1803]: If this function is not called outside `enterprise.ex` make it a private function.

File: lib/wraft_doc/organisation/enterprise.ex [hunk 20]
```diff
@@ -1770,20 +1785,49 @@ defmodule WraftDoc.Enterprise do
...
+    case user.last_signed_in_org == org_id do
```
Comment (vasil1729) [inline lib/wraft_doc/organisation/enterprise.ex:1804]:
```elixir
def handle_last_signed_in_org(%User{last_signed_in_org: last_signed_in_org} = user, org_id) when last_signed_in_org == org_id do
     personal_org = get_personal_organisation_and_role(user)

        user
        |> User.update_last_signed_in_org_changeset(%{
          last_signed_in_org: personal_org.organisation.id
        })
        |> Repo.update()
        |> case do
          {:ok, _updated_user} -> :ok
          error -> error
        end
end

def handle_last_signed_in_org(_, _), do: :ok
```

File: lib/wraft_doc_web/controllers/instance_controller.ex [hunk 21]
```diff
@@ -974,11 +976,21 @@ defmodule WraftDocWeb.Api.V1.InstanceController do
     %{current_org_id: organisation_id} = current_user = conn.assigns.current_user
 
     with %Instance{
-           content_type: %ContentType{organisation_id: ^organisation_id},
-           state: _state
+           content_type: %ContentType{
+             organisation: %Organisation{id: ^organisation_id} = organisation
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/instance_controller.ex:980]: Test for the scenario where you will be hit match error to ensure that error case is handled gracefully by the api. It it not there, add an else block to handle the error.

```elixir
organisation: %Organisation{id: ^organisation_id} = organisation
```

File: lib/wraft_doc_web/controllers/organisation_controller.ex [hunk 22]
```diff
@@ -470,10 +470,11 @@ defmodule WraftDocWeb.Api.V1.OrganisationController do
 
   @spec request_deletion(Plug.Conn.t(), map) :: Plug.Conn.t()
   def request_deletion(conn, _params) do
-    %{current_org_id: organisation_id, email: email} = current_user = conn.assigns.current_user
+    %{current_org_id: organisation_id, email: email, id: user_id} =
+      current_user = conn.assigns.current_user
 
     with {:error, :already_member} <- Enterprise.already_member(organisation_id, email),
-         %Organisation{name: name} = organisation when name != "Personal" <-
+         %Organisation{name: name, owner_id: ^user_id} = organisation when name != "Personal" <-
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/organisation_controller.ex:477]: Test for the scenario where you will be hit match error to ensure that error case is handled gracefully by the api. It it not there, add an else block to handle the error.

```elixir
         %Organisation{name: name, owner_id: ^user_id} = organisation when name != "Personal" <-
```

File: lib/wraft_doc_web/controllers/organisation_controller.ex [hunk 23]
```diff
@@ -642,13 +643,47 @@ defmodule WraftDocWeb.Api.V1.OrganisationController do
...
+  end
+
+  @doc """
+    Transfer Organisation into a new user
+  """
+  @spec transfer_ownership(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def transfer_ownership(conn, %{"id" => new_user_id}) do
+    %{id: user_id} = current_user = conn.assigns[:current_user]
+
+    with %Organisation{owner_id: ^user_id} = organisation <-
```
Comment (vasil1729) [inline lib/wraft_doc_web/controllers/organisation_controller.ex:676]: Test for the scenario where you will be hit match error to ensure that error case is handled gracefully by the api. It it not there, add an else block to handle the error.

line: 676
```elixir
    with %Organisation{owner_id: ^user_id} = organisation <-
```
